crates/matrix-sdk-crypto/src/machine.rs

```diff
@@ -903,6 +904,66 @@ impl OlmMachine {
         self.inner.group_session_manager.invalidate_group_session(room_id).await
     }
 
+    ///
+    /// Creates an encrypted to_device event ready to be sent to that device.
+    /// The caller should ensure that there is an existing Olm session with the
+    /// target device, by calling `get_missing_sessions(user_id)`
+    ///
+    /// # Arguments
+    ///
+    /// * `user_id` - The user ID of the target device.
+    /// * `device_id` - The device ID of the target device.
+    /// * `event_type` - The type of the event to be sent.
+    /// * `content` - The content of the event to be sent. This should be a type
+    ///   that implements the `Serialize` trait.
+    ///
+    /// # Returns
+    /// None if the device is unknown, errors if no olm session is found or
+    /// if an error occurs during encryption.
+    pub async fn create_encrypted_to_device_request(
+        &self,
+        user_id: &UserId,
+        device_id: &DeviceId,
+        event_type: &str,
+        content: &Value,
+    ) -> OlmResult<Option<ToDeviceRequest>> {
+        // Timeout is None since the caller is expected to ensure `get_missing_sessions`
+        // first.
+        let device = self.get_device(user_id, device_id, None).await?;
+
+        if device.is_none() {
+            warn!(
+                "Trying to send an encrypted to-device event to an unknown device {}:{}",
+                user_id, device_id
+            );
+            return Ok(None);
```

Comment:
Why not returning an error here?

Reply:
Rather than returning an error, why not move the method onto the device so you remove the whole branch and you can't call the method if there isn't a device.

Also, you're using an `device.unwrap()` bellow, the idiomatic way to do this would be:

```rust
let Some(device) = device else { return Ok(None); };
```

Reply:
> why not move the method onto the device

I hesitated, if I do that should the method on device returns the changed session for the caller to save? or can the device directly save changes?

Reply:
>  returning an error here

Didn't feel like creating a whole new sort of Result that would have OlmErrors + the unknown device variant

Reply:
> I hesitated, if I do that should the method on device returns the changed session for the caller to save? or can the device directly save changes?

The `Device` should store the `Session`.

Reply:
This would be quite a big change for the bindings? Currently device is just an immutable structure.
Currently for example `verify_device` is on the olmMachine in the bindings.

Or is it just moving the `encrypt_to_device_event` in the device object, but still expose the API in the bindings' OlmMachine? (in that case I'll still need an UnknownDevice error)

Reply:
@poljar I updated the PR and moved the method inside `Device` (it's a bit strange to have to access the store via `verification_machine`). 
For the ffi bindings the method in on the OlmMachine (like verify_device)